62_CAD/shaders/main_pipeline/fragment_shader.hlsl

```diff
@@ -1,655 +1,602 @@
+            // DONE (but needs to be fixed): A. the outlines can be stippled, use phaseshift of the line such that they started from the grid's origin worldspace coordinate
+            // DONE: B. the contours are computed for triangles, use the same function as for dtms, choose between the two triangles based on local UV coords in current cell
+                // DONE: Make it so we can choose which diagonal to use to construct the triangle, it's either u=v or u=1-v
+            // DONE: C. Height shading same as contours (split into two triangles)
+
+            // DONE (but needs to be tested after i implement texture height maps) Heights can have invalid values (let's say NaN) if a cell corner has NaN value then no triangle (for contour and shading) and no outline should include that corner. (see DTM image in discord with gaps)
+            
+            // TODO: we need to emulate dilation and do sdf of neighbouring cells as well. because contours, outlines and shading can bleed into other cells for AA.
+            // [NOTE] Do dilation as last step, when everything else works fine
+
+            DTMSettings dtmSettings = loadDTMSettings(mainObj.dtmSettingsIdx);
+
+            float2 pos = input.getGridDTMScreenSpacePosition();
+            float2 uv = input.getImageUV();
+            const uint32_t textureId = input.getGridDTMHeightTextureID();
+
+            // grid consists of square cells and cells are divided into two triangles:
+            // depending on mode it is
+            // either:        or:
+            // v2a-------v1   v0-------v2b
+            // |  A     / |   | \     B  |
+            // |     /    |   |    \     |
+            // |  /  B    |   |   A   \  |
+            // v0-------v2b   v2a-------v1
+            // 
+
+            // calculate screen space coordinates of vertices of t
+            // he current tiranlge within the grid
+            float3 v[3];
+            nbl::hlsl::shapes::Line<float> outlineLineSegments[2];
+            float outlinePhaseShift;
+            {
+                float2 topLeft = input.getGridDTMScreenSpaceTopLeft();
+                float2 gridExtents = input.getGridDTMScreenSpaceGridExtents();
+                float cellWidth = input.getGridDTMScreenSpaceCellWidth();
+
+                float2 gridSpacePos = uv * gridExtents;
+
+                float2 cellCoords;
+                {
+                    float2 gridSpacePosDivGridCellWidth = gridSpacePos / cellWidth;
+                    cellCoords.x = uint32_t(gridSpacePosDivGridCellWidth.x);
+                    cellCoords.y = uint32_t(gridSpacePosDivGridCellWidth.y);
+                }
+
+                float2 insideCellCoord = gridSpacePos - float2(cellWidth, cellWidth) * cellCoords; // TODO: use fmod instead?
+                
+                const float InvalidHeightValue = asfloat(0x7FC00000);
+                float4 cellHeights = float4(InvalidHeightValue, InvalidHeightValue, InvalidHeightValue, InvalidHeightValue);

```

Comment:
Comment the formation: is it (TL, TR, BL, BR) for example?

Reply:
i don't get it